File: server/src/controllers/market.controller.js
import * as marketService from '../services/market.service.js';

/**
 * GET /timeseries/:symbol
 */
export const getTimeSeries = async (req, res) => {
  try {
    const { symbol } = req.params;
    const data = await marketService.getTimeSeries(symbol);
    res.json(data);
  } catch (error) {
    const status = error.statusCode || 500;
    res.status(status).json({ message: error.message || 'Server error fetching stock data' });
  }
};

/**
 * GET /market-movers
 */
export const getMarketMovers = async (req, res) => {
  try {
    const data = await marketService.getMarketMovers();
    res.json(data);
  } catch (error) {
    const status = error.statusCode || 500;
    res.status(status).json({ message: error.message || 'Server error fetching market movers' });
  }
};

/**
 * GET /quote/:symbol
 */
export const getStockQuote = async (req, res) => {
  try {
    const { symbol } = req.params;
    const data = await marketService.getStockQuote(symbol);
    res.json(data);
  } catch (error) {
    const status = error.statusCode || 500;
    res.status(status).json({ message: error.message || 'Server error fetching stock quote' });
  }
};

/**
 * GET /search?q=AAPL
 */
export const searchSymbol = async (req, res) => {
  try {
    const { q } = req.query;
    if (!q || q.trim().length === 0) {
      return res.json({ data: [] });
    }
    const data = await marketService.searchSymbol(q.trim());
    res.json(data);
  } catch (error) {
    const status = error.statusCode || 500;
    res.status(status).json({ message: error.message || 'Server error searching symbols' });
  }
};

/**
 * GET /logo/:symbol
 */
export const getStockLogo = async (req, res) => {
  try {
    const { symbol } = req.params;
    const data = await marketService.getStockLogo(symbol);
    // TwelveData returns { symbol, logo }, normalize to { symbol, url }
    res.json({ symbol: data.symbol, url: data.logo || data.url || null });
  } catch (error) {
    const status = error.statusCode || 500;
    res.status(status).json({ message: error.message || 'Server error fetching stock logo' });
  }
};

File: corebackend/Controllers/MarketController.cs
using Microsoft.AspNetCore.Mvc;
using corebackend.Services;
using System.Text.Json.Nodes;

namespace corebackend.Controllers
{
    [ApiController]
    [Route("api/v1/market")]
    public class MarketController : ControllerBase
    {
        private readonly TwelveDataService _marketService;

        public MarketController(TwelveDataService marketService)
        {
            _marketService = marketService;
        }

        // GET /api/v1/market/timeseries/:symbol
        [HttpGet("timeseries/{symbol}")]
        public async Task<IActionResult> GetTimeSeries(string symbol)
        {
            try
            {
                var data = await _marketService.GetTimeSeriesAsync(symbol);
                return Ok(data);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = ex.Message });
            }
        }

        // GET /api/v1/market/market-movers
        [HttpGet("market-movers")]
        public async Task<IActionResult> GetMarketMovers()
        {
            try
            {
                var data = await _marketService.GetMarketMoversAsync();
                return Ok(data);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = ex.Message });
            }
        }

        // GET /api/v1/market/quote/:symbol
        [HttpGet("quote/{symbol}")]
        public async Task<IActionResult> GetStockQuote(string symbol)
        {
            try
            {
                var data = await _marketService.GetStockQuoteAsync(symbol);
                return Ok(data);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = ex.Message });
            }
        }

        // GET /api/v1/market/search?q=AAPL
        [HttpGet("search")]
        public async Task<IActionResult> SearchSymbol([FromQuery] string q)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(q))
                {
                    return Ok(new { data = Array.Empty<object>() });
                }

                var data = await _marketService.SearchSymbolAsync(q.Trim());
                return Ok(data);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = ex.Message });
            }
        }

        // GET /api/v1/market/logo/:symbol
        [HttpGet("logo/{symbol}")]
        public async Task<IActionResult> GetStockLogo(string symbol)
        {
            try
            {
                var data = await _marketService.GetStockLogoAsync(symbol);
                
                string? url = null;
                if (data is JsonObject jsonObj)
                {
                    if (jsonObj.TryGetPropertyValue("logo", out var logoNode)) url = logoNode?.ToString();
                    else if (jsonObj.TryGetPropertyValue("url", out var urlNode)) url = urlNode?.ToString();
                }

                return Ok(new { symbol, url });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = ex.Message });
            }
        }
    }
}
